src/include/daos_fs_sys.h

```diff
@@ -0,0 +1,581 @@
+/*
+ * PROPOSAL - Both mpifileutils and daos fs copy have need to interface
+ * with DFS in a manner more similar to POSIX I/O syscalls.
+ * The main difference is that DFS uses object handles to the parent
+ * with a string for just the leaf entry, whereas POSIX uses a string
+ * for the entire path.
+ *
+ * For example, consider the difference between DFS mkdir and POSIX mkdir:
+ *      // DFS version
+ *      int
+ *      dfs_mkdir(dfs_t *dfs, dfs_obj_t *parent, const char *name, mode_t mode,
+ *                daos_oclass_id_t cid);
+ *
+ *      // POSIX version
+ *      int mkdir(const char *pathname, mode_t mode);
+ *
+ * For most of the calls, the boilerplate code of splitting a path into
+ * dirname, basename and calling dfs_lookup on dirname could be abstracted
+ * into a function.
+ *
+ * 1. It would be useful to wrap the dfs_ calls in a posix-equivalent manner,
+ *    including errno, etc.
+ *    Since most of these POSIX calls are #included from sys, we might want
+ *    to use the prefix dfs_sys_ for these wrappers.
+ *
+ * 2. It would be nice if the calls used a d_hash for dfs_lookup calls,
+ *    since the parent directory will likely appear many times.
+ *    Since we need the dfs_t and dfs_obj_t in nearly all cases,
+ *    perhaps we could put a dfs_t, dfs_obj_t, and d_hash in a struct
+ *    that we can pass to each function.
+ * */
+
+/* TODO evaluate includes */
+#include <libgen.h>
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/xattr.h>
+#include <linux/xattr.h>
+#include <daos/checksum.h>
+#include <daos/common.h>
+#include <daos/event.h>
+#include <daos/container.h>
+#include <daos/array.h>
+
+#include "daos.h"
+#include "daos_fs.h"
+
+/** struct holding attributes for the dfs_sys calls */
+typedef struct dfs_sys dfs_sys_t;
+struct dfs_sys {
+	dfs_t			*dfs;		/** mounted filesystem */
+	struct d_hash_table	*dfs_hash;	/** optional lookup hash */
+
+	/* TODO pass obj separately to only the functions that need it? */
+	dfs_obj_t		*obj;		/** the obj */
+};
+
+/** struct holding parsed dirname, name, and cached parent obj */
+typedef struct dfs_sys_path {
```

Comment:
One thing that would work well is if this simply saved two pointers, and two lengths, but into the same allocation, and used the len to control the length of the dir_name string rather than a \0 character.

Reply:
Hmm, I see. I have made a note of this.